sqlglot/dialects/snowflake.py

```diff
@@ -1,11 +1,106 @@
-from sqlglot.dialects.dialect import Dialect
+from sqlglot import exp
+from sqlglot.dialects.dialect import Dialect, format_time_lambda, rename_func
+from sqlglot.expressions import Literal
+from sqlglot.helper import list_get
+from sqlglot.parser import Parser
 from sqlglot.tokens import Tokenizer, TokenType
+from sqlglot.generator import Generator
 
+def _check_int(s):
+    if s[0] in ('-', '+'):
+        return s[1:].isdigit()
+    return s.isdigit()
+
+# from https://docs.snowflake.com/en/sql-reference/functions/to_timestamp.html
+def _snowflake_to_timestamp(args):
+    args_num = len(args)
+    if args_num == 2:
+        first_arg = list_get(args, 0)
+        second_arg = list_get(args, 1)
+        if first_arg.is_string:
+            # case: <string_expr> [ , <format> ]
+            return format_time_lambda(exp.StrToTime, "snowflake")
+
+        # case: <numeric_expr> [ , <scale> ]
+        if second_arg.this not in ['0', '3', '9']:
+            raise ValueError(f"Scale for snowflake numeric timestamp is {second_arg}, but should be 0, 3, or 9")
+
+        def _convert_time_scale_and_run(args):
+            conv_args = []
+            exponent = int(list_get(args, 1).this)
+            retval = int(int(list_get(args, 0).this) / (10 ** exponent))
+            conv_args.append(str(retval))
+            args[:] = args[:1]
```

Comment:
what are you doing with args?

Reply:
If you look at the possible ways TO_TIMESTAMP can be used in Snowflake (https://docs.snowflake.com/en/sql-reference/functions/to_timestamp.html#arguments), one possible option is that it takes a Unix timestamp and a scale argument to decide whether to treat that timestamp as seconds, milliseconds, or nanoseconds.

If that scale option is specified, we need to divide the timestamp by `10**scale`, so we can use the timestamp as an argument to UnixToTime which only takes one argument (the timestamp itself). After division we replace the (timestamp, scale) arguments with just the adjusted timestamp.

Reply:
but you don't use args, you pass in conv_args?

Reply:
Unfortunately it doesn't seem like I can modify the value of `args[0]` directly, as `args[0].this` is protected by an `@property`. When we create the `UnixToTime` instance we pass `conv_args`, and the SQL is parsed correctly as a result, but when we then call the function in the lambda in `FUNCTIONS`, that requires only one argument rather than two, so we need to truncate `args` to get that function to work.

Reply:
i don't understand what mutating args does, do you use it anywhere, specifically args[:] = args[:1]

can you just do

return exp.UnixTimeToTime(this=str(int(x) / (10 * int(y))

Reply:
I cleaned up the code a little bit. The issue is that we get an error from `validate_expression` instantiating `UnixToTime` because `args` itself is length 2 rather than 1:

```
sqlglot.errors.ParseError: The number of provided arguments (2) is greater than the maximum number of supported arguments (1). Line 1, Col: 37.
  SELECT TO_TIMESTAMP(1659981729000, 3)
```

The truncation of `args` is to avoid this error.

Reply:
One way to avoid this is modifying the `UnixToTime` class in `expressions.py` from

```python
class UnixToTime(Func):
    pass
```

to

```python
class UnixToTime(Func):
    arg_types = {"this": True, "scale": False}
```

(i.e. a non-required "scale" argument). If this is preferable let me know and I'll make the change.

Reply:
yep that makes sense to me

Reply:
fixed